tools/c7n_mailer/c7n_mailer/splunk_delivery.py

```diff
@@ -0,0 +1,279 @@
+                if maxlen is not None and len(p) > maxlen:
+                    # never retry if we're over configured max length
+                    return False
+                return True  # if no exception, just return
+            except Exception:
+                sleep_sec = uniform(1, 4)  # random float 1 to 4
+                self.logger.warning(
+                    'Caught exception sending to Splunk; '
+                    'retry in %s seconds', sleep_sec
+                )
+                sleep(sleep_sec)
+        self.logger.error(
+            'ERROR - Could not POST to Splunk after %d tries.', max_attempts
+        )
+        return False
+
+    def _send_splunk(self, payload):
+        """
+        Perform the actual data send to Splunk HEC for one log entry
+
+        :param payload: the JSON-encoded payload to send to Splunk as
+          event data
+        :type payload: str
+        """
+        url = self.config['splunk_hec_url']
+        self.logger.debug('Send to Splunk (%s): %s', url, payload)
+        try:
+            r = requests.post(
+                url,
+                headers={
+                    'Authorization': 'Splunk %s' % self.config[
+                        'splunk_hec_token'
+                    ]
+                },
+                data=payload
+            )
+        except Exception:
+            self.logger.error('Exception during Splunk POST to %s of %s',
+                              url, payload, exc_info=True)
+            raise
+        self.logger.debug(
+            'Splunk POST got response code %s HEADERS=%s BODY: %s',
+            r.status_code, r.headers, r.text
+        )
+        if r.status_code not in [200, 201, 202]:
+            self.logger.error(
+                'Splunk POST returned non-20x response: %s HEADERS=%s BODY: %s',
+                r.status_code, r.headers, r.text
+            )
+            raise RuntimeError('POST returned %s' % r.status_code)
+        try:
+            j = r.json()
+        except Exception:
+            j = {'text': r.text}
+        if j['text'].lower() != 'success':
+            self.logger.error(
+                'Splunk POST returned non-success response: %s', j
+            )
+            raise RuntimeError('POST returned non-success response: %s' % j)
+
+    @staticmethod
+    def tags_for_resource(res):
+        """
+        Return the tags for a given resource, or an empty dict if they
+        can't be found.
+
+        :param res: c7n resource info
+        :type res: dict
+        :return: dict of tags
+        """
+        try:
+            t = res.get('Tags', res.get('tags', {}))
+            if t is None:
+                return {}
+            # it's a dict, return it right away
+            if isinstance(t, type({})):
+                return t
+            # if it's not a dict or list, unknown, return empty dict
+            if not isinstance(t, type([])):
+                return {}
+            # it's a list
```

Comment:
is there any resource you've seen where this isn't the case? the tag: aliasing for value filter is intended to work with list form, which we normalize resources to.

Reply:
This may be a relic of issues that we saw over a year ago (when the code was originally seen), but at that time at least, we'd had issues with Event or CloudTrail-based policies triggering on resources that had been deleted before the policy itself ran, and ending up with... None, maybe? Though it's entirely possible this logic is no longer needed if resources are always normalized to a list.

Reply:
After going through ``resources.json.gz`` for every execution of our policies so far in 2019 (except some cloudtrail mode policies, which for some reason aren't writing logs to S3), I can confirm that for every resource matched by all of them, either there's a ``Tags`` key with a value that's a list of dicts with at least ``Key`` and ``Value`` elements, or the ``Tags`` key is missing and there are no tags.

I've pushed a commit that drastically simplifies the tag dict logic.